File: carRental.py
class Car:
    def _init_(self, carId, brand, model, basePricePerDay):
        self.carId = carId
        self.brand = brand
        self.model = model
        self.basePricePerDay = basePricePerDay
        self.available = True

    def getCarId(self):
        return self.carId

    def getBrand(self):
        return self.brand

    def getModel(self):
        return self.model

    def calculatePrice(self, rentalDays):
        return self.basePricePerDay * rentalDays

    def isCarAvailable(self):
        return self.available

    def rent(self):
        self.available = False

    def returnCar(self):
        self.available = True


class Customer:
    def _init_(self, customerId, name):
        self.customerId = customerId
        self.name = name

    def getCustomerId(self):
        return self.customerId

    def getName(self):
        return self.name


class Rental:
    def _init_(self, car, customer, days):
        self.car = car
        self.customer = customer
        self.days = days

    def getCar(self):
        return self.car

    def getCustomer(self):
        return self.customer

    def getDays(self):
        return self.days


class CarRentalSystem:
    def _init_(self):
        self.cars = []
        self.customers = []
        self.rentals = []

        # Adding initial cars
        car1 = Car("CAR1", "Toyota", "Corolla", 30.0)
        car2 = Car("CAR2", "Honda", "Civic", 35.0)
        car3 = Car("CAR3", "Ford", "Focus", 40.0)
        car4 = Car("CAR4", "Chevrolet", "Malibu", 32.5)
        car5 = Car("CAR5", "Nissan", "Altima", 33.8)
        car6 = Car("CAR6", "BMW", "3 Series", 37.2)
        car7 = Car("CAR7", "Mercedes-Benz", "C-Class", 36.5)
        car8 = Car("CAR8", "Audi", "A4", 34.7)
        car9 = Car("CAR9", "Hyundai", "Elantra", 31.5)
        car10 = Car("CAR10", "Kia", "Forte", 29.3)
        car11 = Car("CAR11", "Mazda", "Mazda3", 30.9)
        car12 = Car("CAR12", "Volkswagen", "Jetta", 28.6)
        car13 = Car("CAR13", "Subaru", "Impreza", 30.2)
        car14 = Car("CAR14", "Chrysler", "300", 33.4)
        car15 = Car("CAR15", "Buick", "Regal", 31.8)
        car16 = Car("CAR16", "Lexus", "IS", 36.1)
        car17 = Car("CAR17", "Infiniti", "Q50", 35.5)
        car18 = Car("CAR18", "Acura", "ILX", 34.0)
        car19 = Car("CAR19", "Genesis", "G70", 38.2)
        car20 = Car("CAR20", "Jaguar", "XE", 39.6)

        self.cars.append(car1)
        self.cars.append(car2)
        self.cars.append(car3)
        self.cars.append(car4)
        self.cars.append(car5)
        self.cars.append(car6)
        self.cars.append(car7)
        self.cars.append(car8)
        self.cars.append(car9)
        self.cars.append(car10)
        self.cars.append(car11)
        self.cars.append(car12)
        self.cars.append(car13)
        self.cars.append(car14)
        self.cars.append(car15)
        self.cars.append(car16)
        self.cars.append(car17)
        self.cars.append(car18)
        self.cars.append(car19)
        self.cars.append(car20)

    def addCar(self, car):
        self.cars.append(car)

    def addCustomer(self, customer):
        self.customers.append(customer)

    def rentCar(self, car, customer, days):
        if car.isCarAvailable():
            car.rent()
            self.rentals.append(Rental(car, customer, days))
        else:
            print("Car is not available for rent.")

    def returnCar(self, car):
        car.returnCar()
        rentalToRemove = None
        for rental in self.rentals:
            if rental.getCar() == car:
                rentalToRemove = rental
                break
        if rentalToRemove is not None:
            self.rentals.remove(rentalToRemove)
        else:
            print("Car was not rented.")

    def getCarById(self, carId):
        for car in self.cars:
            if car.getCarId() == carId:
                return car
        return None

    def getRentalByCarId(self, carId):
        for rental in self.rentals:
            if rental.getCar().getCarId() == carId:
                return rental
        return None

    def addNewCar(self, car):
        self.cars.append(car)

    def removeCar(self, carId):
        carToRemove = self.getCarById(carId)
        rentalToRemove = self.getRentalByCarId(carId)

        if carToRemove is not None:
            if rentalToRemove is not None:
                print("Car is currently rented. Cannot remove.")
                return
            self.cars.remove(carToRemove)
            print("Car removed successfully.")
        else:
            print("Car not found.")

    def menu(self):
        while True:
            print("1. Rent a Car")
            print("2. Return a Car")
            print("3. Add a Car")
            print("4. Remove a Car")
            print("5. Exit")
            choice = int(input("Enter your choice: "))

            if choice == 1:
                print("== Rent a Car ==")
                customerName = input("Enter your name: ")

                print("Available Cars:")
                for car in self.cars:
                    if car.isCarAvailable():
                        print(car.getCarId(), "-", car.getBrand(), car.getModel())

                carId = input("Enter the car ID you want to rent: ")
                rentalDays = int(input("Enter the number of days for rental: "))

                newCustomer = Customer("CUS" + str(len(self.customers) + 1), customerName)
                self.addCustomer(newCustomer)

                selectedCar = None
                for car in self.cars:
                    if car.getCarId() == carId and car.isCarAvailable():
                        selectedCar = car
                        break

                if selectedCar is not None:
                    totalPrice = selectedCar.calculatePrice(rentalDays)
                    print("== Rental Information ==")
                    print("Customer ID:", newCustomer.getCustomerId())
                    print("Customer Name:", newCustomer.getName())
                    print("Car:", selectedCar.getBrand(), selectedCar.getModel())
                    print("Rental Days:", rentalDays)
                    print("Total Price: $", format(totalPrice, ".2f"))

                    confirm = input("Confirm rental (Y/N): ")

                    if confirm == "Y" or confirm == "y":
                        self.rentCar(selectedCar, newCustomer, rentalDays)
                        print("Car rented successfully.")
                    else:
                        print("Rental canceled.")
                else:
                    print("Invalid car selection or car not available for rent.")
            elif choice == 2:
                print("== Return a Car ==")
                carId = input("Enter the car ID you want to return: ")

                carToReturn = None
                for car in self.cars:
                    if car.getCarId() == carId and not car.isCarAvailable():
                        carToReturn = car
                        break

                if carToReturn is not None:
                    customer = None
                    for rental in self.rentals:
                        if rental.getCar() == carToReturn:
                            customer = rental.getCustomer()
                            break

                    if customer is not None:
                        self.returnCar(carToReturn)
                        print("Car returned successfully by", customer.getName())
                    else:
                        print("Car was not rented or rental information is missing.")
                else:
                    print("Invalid car ID or car is not rented.")
            elif choice == 3:
                print("== Add a Car ==")
                carId = input("Enter the car ID: ")
                brand = input("Enter the car brand: ")
                model = input("Enter the car model: ")
                basePricePerDay = float(input("Enter the car's base price per day: "))

                newCar = Car(carId, brand, model, basePricePerDay)
                self.addCar(newCar)

                print("Car added successfully.")
            elif choice == 4:
                print("== Remove a Car ==")
                carId = input("Enter the car ID: ")

                self.removeCar(carId)
            elif choice == 5:
                break
            else:
                print("Invalid choice. Please enter a valid option.")

        print("Thank you for using the Car Rental System!")


carRentalSystem = CarRentalSystem()
carRentalSystem.menu()

File: carRental.cpp
#include <iostream>
#include <vector>
#include <string>
#include <algorithm>
#include <iomanip> // for std::setprecision

class Car {
private:
    std::string carId;
    std::string brand;
    std::string model;
    double basePricePerDay;
    bool available;

public:
    Car(std::string carId, std::string brand, std::string model, double basePricePerDay) {
        this->carId = carId;
        this->brand = brand;
        this->model = model;
        this->basePricePerDay = basePricePerDay;
        this->available = true;
    }

    std::string getCarId() {
        return carId;
    }

    std::string getBrand() {
        return brand;
    }

    std::string getModel() {
        return model;
    }

    double calculatePrice(int rentalDays) {
        return basePricePerDay * rentalDays;
    }

    bool isCarAvailable() {
        return available;
    }

    void rent() {
        available = false;
    }

    void returnCar() {
        available = true;
    }
};

class Customer {
private:
    std::string customerId;
    std::string name;

public:
    Customer(std::string customerId, std::string name) {
        this->customerId = customerId;
        this->name = name;
    }

    std::string getCustomerId() {
        return customerId;
    }

    std::string getName() {
        return name;
    }
};

class Rental {
private:
    Car* car;
    Customer* customer;
    int days;

public:
    Rental(Car* car, Customer* customer, int days) {
        this->car = car;
        this->customer = customer;
        this->days = days;
    }

    Car* getCar() {
        return car;
    }

    Customer* getCustomer() {
        return customer;
    }

    int getDays() {
        return days;
    }
};

class CarRentalSystem {
private:
    std::vector<Car*> cars;
    std::vector<Customer*> customers;
    std::vector<Rental*> rentals;

public:
    CarRentalSystem() {
        // Adding initial cars
        cars.push_back(new Car("CAR1", "Toyota", "Corolla", 30.0));
        cars.push_back(new Car("CAR2", "Honda", "Civic", 35.0));
        cars.push_back(new Car("CAR3", "Ford", "Focus", 40.0));
        cars.push_back(new Car("CAR4", "Chevrolet", "Malibu", 32.5));
        cars.push_back(new Car("CAR5", "Nissan", "Altima", 33.8));
        cars.push_back(new Car("CAR6", "BMW", "3 Series", 37.2));
        cars.push_back(new Car("CAR7", "Mercedes-Benz", "C-Class", 36.5));
        cars.push_back(new Car("CAR8", "Audi", "A4", 34.7));
        cars.push_back(new Car("CAR9", "Hyundai", "Elantra", 31.5));
        cars.push_back(new Car("CAR10", "Kia", "Forte", 29.3));
        cars.push_back(new Car("CAR11", "Mazda", "Mazda3", 30.9));
        cars.push_back(new Car("CAR12", "Volkswagen", "Jetta", 28.6));
        cars.push_back(new Car("CAR13", "Subaru", "Impreza", 30.2));
        cars.push_back(new Car("CAR14", "Chrysler", "300", 33.4));
        cars.push_back(new Car("CAR15", "Buick", "Regal", 31.8));
        cars.push_back(new Car("CAR16", "Lexus", "IS", 36.1));
        cars.push_back(new Car("CAR17", "Infiniti", "Q50", 35.5));
        cars.push_back(new Car("CAR18", "Acura", "ILX", 34.0));
        cars.push_back(new Car("CAR19", "Genesis", "G70", 38.2));
        cars.push_back(new Car("CAR20", "Jaguar", "XE", 39.6));


    }

    void addCar(Car* car) {
        cars.push_back(car);
    }

    void addCustomer(Customer* customer) {
        customers.push_back(customer);
    }

    void rentCar(Car* car, Customer* customer, int days) {
        if (car->isCarAvailable()) {
            car->rent();
            rentals.push_back(new Rental(car, customer, days));
        } else {
            std::cout << "Car is not available for rent." << std::endl;
        }
    }

    void returnCar(Car* car) {
        car->returnCar();
        Rental* rentalToRemove = nullptr;
        for (Rental* rental : rentals) {
            if (rental->getCar() == car) {
                rentalToRemove = rental;
                break;
            }
        }
        if (rentalToRemove != nullptr) {
            rentals.erase(std::remove(rentals.begin(), rentals.end(), rentalToRemove), rentals.end());
        } else {
            std::cout << "Car was not rented." << std::endl;
        }
    }

    Car* getCarById(std::string carId) {
        for (Car* car : cars) {
            if (car->getCarId() == carId) {
                return car;
            }
        }
        return nullptr;
    }

    Rental* getRentalByCarId(std::string carId) {
        for (Rental* rental : rentals) {
            if (rental->getCar()->getCarId() == carId) {
                return rental;
            }
        }
        return nullptr;
    }

    void addNewCar(Car* car) {
        cars.push_back(car);
    }

    void removeCar(std::string carId) {
        Car* carToRemove = getCarById(carId);
        Rental* rentalToRemove = getRentalByCarId(carId);

        if (carToRemove != nullptr) {
            if (rentalToRemove != nullptr) {
                std::cout << "Car is currently rented. Cannot remove." << std::endl;
                return;
            }
            cars.erase(std::remove(cars.begin(), cars.end(), carToRemove), cars.end());
            std::cout << "Car removed successfully." << std::endl;
        } else {
            std::cout << "Car not found." << std::endl;
        }
    }

    void menu() {
        while (true) {
            std::cout << "1. Rent a Car" << std::endl;
            std::cout << "2. Return a Car" << std::endl;
            std::cout << "3. Add a Car" << std::endl;
            std::cout << "4. Remove a Car" << std::endl;
            std::cout << "5. Exit" << std::endl;
            std::cout << "Enter your choice: ";

            int choice;
            std::cin >> choice;
            std::cin.ignore(); // Consume newline

            if (choice == 1) {
                std::cout << "== Rent a Car ==" << std::endl;
                std::cout << "Enter your name: ";
                std::string customerName;
                std::getline(std::cin, customerName);

                std::cout << "Available Cars:" << std::endl;
                for (Car* car : cars) {
                    if (car->isCarAvailable()) {
                        std::cout << car->getCarId() << " - " << car->getBrand() << " " << car->getModel() << std::endl;
                    }
                }

                std::cout << "Enter the car ID you want to rent: ";
                std::string carId;
                std::getline(std::cin, carId);

                std::cout << "Enter the number of days for rental: ";
                int rentalDays;
                std::cin >> rentalDays;
                std::cin.ignore(); // Consume newline

                Customer* newCustomer = new Customer("CUS" + std::to_string(customers.size() + 1), customerName);
                addCustomer(newCustomer);

                Car* selectedCar = nullptr;
                for (Car* car : cars) {
                    if (car->getCarId() == carId && car->isCarAvailable()) {
                        selectedCar = car;
                        break;
                    }
                }

                if (selectedCar != nullptr) {
                    double totalPrice = selectedCar->calculatePrice(rentalDays);
                    std::cout << "== Rental Information ==" << std::endl;
                    std::cout << "Customer ID: " << newCustomer->getCustomerId() << std::endl;
                    std::cout << "Customer Name: " << newCustomer->getName() << std::endl;
                    std::cout << "Car: " << selectedCar->getBrand() << " " << selectedCar->getModel() << std::endl;
                    std::cout << "Rental Days: " << rentalDays << std::endl;
                    std::cout << "Total Price: $" << std::fixed << std::setprecision(2) << totalPrice << std::endl;

                    std::cout << "Confirm rental (Y/N): ";
                    std::string confirm;
                    std::getline(std::cin, confirm);

                    if (confirm == "Y" || confirm == "y") {
                        rentCar(selectedCar, newCustomer, rentalDays);
                        std::cout << "Car rented successfully." << std::endl;
                    } else {
                        std::cout << "Rental canceled." << std::endl;
                    }
                } else {
                    std::cout << "Invalid car selection or car not available for rent." << std::endl;
                }
            } else if (choice == 2) {
                std::cout << "== Return a Car ==" << std::endl;
                std::cout << "Enter the car ID you want to return: ";
                std::string carId;
                std::getline(std::cin, carId);

                Car* carToReturn = nullptr;
                for (Car* car : cars) {
                    if (car->getCarId() == carId && !car->isCarAvailable()) {
                        carToReturn = car;
                        break;
                    }
                }

                if (carToReturn != nullptr) {
                    Customer* customer = nullptr;
                    for (Rental* rental : rentals) {
                        if (rental->getCar() == carToReturn) {
                            customer = rental->getCustomer();
                            break;
                        }
                    }

                    if (customer != nullptr) {
                        returnCar(carToReturn);
                        std::cout << "Car returned successfully by " << customer->getName() << std::endl;
                    } else {
                        std::cout << "Car was not rented or rental information is missing." << std::endl;
                    }
                } else {
                    std::cout << "Invalid car ID or car is not rented." << std::endl;
                }
            } else if (choice == 3) {
                std::cout << "== Add a Car ==" << std::endl;
                std::cout << "Enter the car ID: ";
                std::string carId;
                std::getline(std::cin, carId);

                std::cout << "Enter the car brand: ";
                std::string brand;
                std::getline(std::cin, brand);

                std::cout << "Enter the car model: ";
                std::string model;
                std::getline(std::cin, model);

                std::cout << "Enter the car's base price per day: ";
                double basePricePerDay;
                std::cin >> basePricePerDay;
                std::cin.ignore(); // Consume newline

                Car* newCar = new Car(carId, brand, model, basePricePerDay);
                addCar(newCar);

                std::cout << "Car added successfully." << std::endl;
            } else if (choice == 4) {
                std::cout << "== Remove a Car ==" << std::endl;
                std::cout << "Enter the car ID: ";
                std::string carId;
                std::getline(std::cin, carId);

                removeCar(carId);
            } else if (choice == 5) {
                break;
            } else {
                std::cout << "Invalid choice. Please enter a valid option." << std::endl;
            }
        }

        std::cout << "Thank you for using the Car Rental System!" << std::endl;
    }
};

int main() {
    CarRentalSystem carRentalSystem;

    carRentalSystem.menu();

    return 0;
}